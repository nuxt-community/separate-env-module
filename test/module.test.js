jest.setTimeout(60000)

const { Nuxt, Builder } = require('nuxt-edge')
const consola = require('consola')
const getPort = require('get-port')

let nuxt, port

const url = path => `http://localhost:${port}${path}`

const setupNuxt = async (config) => {
  config.dev = true
  const nuxt = new Nuxt(config)
  await new Builder(nuxt).build()
  port = await getPort()
  await nuxt.listen(port)

  return nuxt
}

const renderSpaRoute = async (nuxt, _url) => {
  const window = await nuxt.renderAndGetWindow(url(_url))
  const head = window.document.head.innerHTML
  const html = window.document.body.innerHTML
  return { window, head, html }
}

describe('csr', () => {
  test('only client-side variables are available', async () => {
    nuxt = await setupNuxt(require('./fixture/configs/csr'))
    const { html } = await renderSpaRoute(nuxt, '/')
    expect(html).toMatchSnapshot()
  })

  afterEach(async () => {
    if (nuxt) {
      await nuxt.close()
    }
  })
})

describe('no object', () => {
  let log

  beforeEach(() => {
    log = jest.fn()
    consola.clear().add({ log })
  })

  test('error when no object is provided', async () => {
    nuxt = await setupNuxt(require('./fixture/configs/no-object'))
    const expectedMessage = 'Could not assign server env variables. Please provide an object or a function that returns an object!'
    const consolaMessages = log.mock.calls.map(c => c[0].message)
    expect(consolaMessages).toContain(expectedMessage)
  })

  afterEach(async () => {
    if (nuxt) {
      await nuxt.close()
    }
  })
})

describe('ssr', () => {
  test('correctly assign env variables from function', async () => {
    const nuxt = await setupNuxt(require('./fixture/configs/ssr-with-function'))
    const { html } = await nuxt.renderRoute('/')
    expect(html).toMatchSnapshot()
  })

  test('server variables only accessible on server-bundle', async () => {
    nuxt = await setupNuxt(require('./fixture/configs/ssr'))
    const { html } = await nuxt.renderRoute('/')
    expect(html).toMatchSnapshot()
  })

  test('honor custom build.extend function', async () => {
    nuxt = await setupNuxt(require('./fixture/configs/ssr-with-extend'))
    const { html } = await nuxt.renderRoute('/')
    expect(html).toMatchSnapshot()
  })

  afterEach(async () => {
    if (nuxt) {
      await nuxt.close()
    }
  })
})
