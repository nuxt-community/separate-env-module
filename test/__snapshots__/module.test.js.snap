// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csr only client-side variables are available 1`] = `
"
    <div id=\\"__nuxt\\"><!----><div id=\\"__layout\\"><div id=\\"result\\" class=\\"appear appear-active\\"><h1>client</h1> <pre>{\\"ONLY_CLIENT\\":\\"okay\\",\\"ONLY_SERVER\\":\\"undefined\\",\\"DIFFERENT\\":\\"client\\",\\"normal\\":\\"Hi\\",\\"number\\":\\"undefined\\"}</pre></div></div></div>
  <script type=\\"text/javascript\\" src=\\"/_nuxt/runtime.js\\"></script><script type=\\"text/javascript\\" src=\\"/_nuxt/commons.app.js\\"></script><script type=\\"text/javascript\\" src=\\"/_nuxt/vendors.app.js\\"></script><script type=\\"text/javascript\\" src=\\"/_nuxt/app.js\\"></script>

"
`;

exports[`ssr correctly assign env variables from function 1`] = `
"<!doctype html>
<html data-n-head-ssr data-n-head=\\"\\">
  <head data-n-head=\\"\\">
    <title data-n-head=\\"true\\"></title><style data-vue-ssr-id=\\"9de64dc4:0\\">
.nuxt-progress {
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  height: 2px;
  width: 0%;
  opacity: 1;
  transition: width 0.1s, opacity 0.4s;
  background-color: black;
  z-index: 999999;
}
.nuxt-progress.nuxt-progress-notransition {
  transition: none;
}
.nuxt-progress-failed {
  background-color: red;
}
</style>
  </head>
  <body data-n-head=\\"\\">
    <div data-server-rendered=\\"true\\" id=\\"__nuxt\\"><!----><div id=\\"__layout\\"><div id=\\"result\\"><h1>server</h1> <pre>{&quot;ONLY_CLIENT&quot;:&quot;undefined&quot;,&quot;ONLY_SERVER&quot;:&quot;yup&quot;,&quot;DIFFERENT&quot;:&quot;server&quot;,&quot;normal&quot;:&quot;Hi&quot;,&quot;number&quot;:&quot;undefined&quot;}</pre></div></div></div><script>window.__NUXT__=(function(a,b){return {layout:\\"default\\",data:[{type:a,env:{ONLY_CLIENT:b,ONLY_SERVER:\\"yup\\",DIFFERENT:a,normal:\\"Hi\\",number:b}}],error:null,serverRendered:true}}(\\"server\\",\\"undefined\\"));</script><script src=\\"/_nuxt/runtime.js\\" defer></script><script src=\\"/_nuxt/pages/index.js\\" defer></script><script src=\\"/_nuxt/commons.app.js\\" defer></script><script src=\\"/_nuxt/vendors.app.js\\" defer></script><script src=\\"/_nuxt/app.js\\" defer></script>
  </body>
</html>
"
`;

exports[`ssr honor custom build.extend function 1`] = `
"<!doctype html>
<html data-n-head-ssr data-n-head=\\"\\">
  <head data-n-head=\\"\\">
    <title data-n-head=\\"true\\"></title><style data-vue-ssr-id=\\"9de64dc4:0\\">
.nuxt-progress {
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  height: 2px;
  width: 0%;
  opacity: 1;
  transition: width 0.1s, opacity 0.4s;
  background-color: black;
  z-index: 999999;
}
.nuxt-progress.nuxt-progress-notransition {
  transition: none;
}
.nuxt-progress-failed {
  background-color: red;
}
</style>
  </head>
  <body data-n-head=\\"\\">
    <div data-server-rendered=\\"true\\" id=\\"__nuxt\\"><!----><div id=\\"__layout\\"><div id=\\"result\\"><h1>server</h1> <pre>{&quot;ONLY_CLIENT&quot;:&quot;undefined&quot;,&quot;ONLY_SERVER&quot;:&quot;nope&quot;,&quot;DIFFERENT&quot;:&quot;server&quot;,&quot;normal&quot;:&quot;Hi&quot;,&quot;number&quot;:&quot;undefined&quot;}</pre></div></div></div><script>window.__NUXT__=(function(a,b){return {layout:\\"default\\",data:[{type:a,env:{ONLY_CLIENT:b,ONLY_SERVER:\\"nope\\",DIFFERENT:a,normal:\\"Hi\\",number:b}}],error:null,serverRendered:true}}(\\"server\\",\\"undefined\\"));</script><script src=\\"/_nuxt/runtime.js\\" defer></script><script src=\\"/_nuxt/pages/index.js\\" defer></script><script src=\\"/_nuxt/commons.app.js\\" defer></script><script src=\\"/_nuxt/vendors.app.js\\" defer></script><script src=\\"/_nuxt/app.js\\" defer></script>
  </body>
</html>
"
`;

exports[`ssr server variables only accessible on server-bundle 1`] = `
"<!doctype html>
<html data-n-head-ssr data-n-head=\\"\\">
  <head data-n-head=\\"\\">
    <title data-n-head=\\"true\\"></title><style data-vue-ssr-id=\\"9de64dc4:0\\">
.nuxt-progress {
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  height: 2px;
  width: 0%;
  opacity: 1;
  transition: width 0.1s, opacity 0.4s;
  background-color: black;
  z-index: 999999;
}
.nuxt-progress.nuxt-progress-notransition {
  transition: none;
}
.nuxt-progress-failed {
  background-color: red;
}
</style>
  </head>
  <body data-n-head=\\"\\">
    <div data-server-rendered=\\"true\\" id=\\"__nuxt\\"><!----><div id=\\"__layout\\"><div id=\\"result\\"><h1>server</h1> <pre>{&quot;ONLY_CLIENT&quot;:&quot;undefined&quot;,&quot;ONLY_SERVER&quot;:&quot;yup&quot;,&quot;DIFFERENT&quot;:&quot;server&quot;,&quot;normal&quot;:&quot;Hi&quot;,&quot;number&quot;:10}</pre></div></div></div><script>window.__NUXT__=(function(a){return {layout:\\"default\\",data:[{type:a,env:{ONLY_CLIENT:\\"undefined\\",ONLY_SERVER:\\"yup\\",DIFFERENT:a,normal:\\"Hi\\",number:10}}],error:null,serverRendered:true}}(\\"server\\"));</script><script src=\\"/_nuxt/runtime.js\\" defer></script><script src=\\"/_nuxt/pages/index.js\\" defer></script><script src=\\"/_nuxt/commons.app.js\\" defer></script><script src=\\"/_nuxt/vendors.app.js\\" defer></script><script src=\\"/_nuxt/app.js\\" defer></script>
  </body>
</html>
"
`;
